docs/decisions/0015-serving-static-assets.rst

```diff
@@ -0,0 +1,165 @@
+
+The assets will be served from a completely different domain from the LMS and Studio, and will not be a subdomain.
+
+A more concrete example: ``https://studio.assets.sandbox.openedx.io/apps/content_libraries/lib:Axim:200/xblock.v1:problem@826eb471-0db2-4943-b343-afa65a6fdeb5/v2/static/images/fig1.png``
+
+The ``version`` can be:
+
+* ``draft`` indicating the latest draft version (viewed by authors in Studio).
+* ``published`` indicating the latest published version (viewed by students in the LMS)
+* ``v{num}`` meaning a specific version–e.g. ``v20`` for version 20.
+
+Asset Server Implementation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There will be two asset server URLs–one corresponding to the LMS and one corresponding to Studio, each with their own subdomain. An example set of domains might be:
+
+* LMS: ``sandbox.openedx.org``
+* Studio: ``studio.sandbox.openedx.org``
+* LMS Assets: ``lms.assets.sandbox.openedx.io`` (note the ``.io`` top level domain)
+* Studio Assets: ``studio.assets.sandbox.openedx.io``
+
+The asset serving domains will be serviced by a Caddy instance that is configured as a reverse proxy to the LMS or Studio. Caddy will be configured to only proxy a specific set of paths that correspond to valid asset URLs.
+
+Django View Implemenation
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The LMS and Studio will each have one or two apps that implement view endpoints by extending a view that will be provided by the Learning Core. These views will only respond to requests that come via the asset domains (i.e. they will not work if you request the same paths using the LMS or Studio domains).
+
+Django is poorly suited to serving large static assets, particularly when deployed using WSGI. Instead of streaming the actual file data, the Django views serving assets will make use of the ``X-Accel-Redirect`` header. This header is supported by both Caddy and Nginx, and will cause them to fetch the data from the specified URI to send to the user. This redirect happens internally in the proxy and does *not* change the browser address. In most cases, the Django view will send a signed URL to a file in an object store.
+
+The Django view will also be responsible for setting other important header information, such as size, content type, and caching information.
+
+Object Store Requirement
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+**This is the first major piece of courseware functionality that would *require* an object store to work correctly.** It does not require that you configure *all* your media files to go to an object store, but it would require that all new assets in this system go there.
+
+It is possible to add support for a file-backed storages option for this, but there are major drawbacks:
+
+#. To be secure, it would either require sharing the asset raw data files between Studio/LMS and the asset server containers, or it would require serving the assets through Django itself (significantly adding to server load).
```

Comment:
> between Studio/LMS and the asset server containers

Is this out of date? Since above it says that the "asset server" is just a different domain name that is served by the LMS/Studio (in the same container).

> it would require serving the assets through Django itself

Doesn't X-accel-redirect solve this problem? You can redirect to a static file on the filesystem, in the media folder.

Reply:
Actually, maybe this is just me not understanding how we store media files without Minio today. Do you know how that works?  Where do those files actually live, and how does Studio/LMS access them?

Reply:
On tutor, they're just files in `/openedx/media` and they're served [by uwsgi](https://github.com/overhangio/tutor/blob/10c988127df5bccc37f9b769088bac0c8ebdcff7/tutor/templates/build/openedx/settings/uwsgi.ini#L1-L3).